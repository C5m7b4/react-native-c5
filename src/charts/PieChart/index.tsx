import React from 'react';
import { View, Dimensions } from 'react-native';
import Svg, { Circle, G, Line, Text as SvgText } from 'react-native-svg';
import { PieProps } from '../../interfaces';

const { width: SCREEN_WIDTH } = Dimensions.get('window');

const PieChart = <T,>(props: PieProps<T>) => {
  const {
    data = [],
    label,
    value,
    height: containerHeight = 300,
    width: containerWidth = SCREEN_WIDTH - 50,
    fontSize = 12,
    textColor = '#000',
    pieBackgroundColor = 'transparent',
    chartBackgroundColor = 'transparent',
    useDonut = true,
    donutRadius = containerHeight / 6,
    donutColor = '#fff',
    tickColor = '#000',
    tickTextColor = '#000',
    colors = [
      '#fcba03',
      '#27a34c',
      '#36b3c9',
      '#3f27b8',
      '#a937cc',
      '#c71e61',
      '#478713',
      '#6b83bf',
    ],
  } = props;
  const padding = 25;
  const centerX = containerWidth / 2;
  const centerY = containerHeight / 2;
  let radius = centerY - padding - 10;
  const smallerRadius = radius / 2;
  const circumference = 2 * Math.PI * smallerRadius;

  const getPiePercent = (percent: number) => {
    const result = (percent * circumference) / 100;
    // const radians = (percent * (2 * Math.PI)) / 360;
    // const degrees = (radians * 360) / (2 * Math.PI);
    // const lineAngle = 360 / (100 / percent);

    return result;
  };

  const calculateLineAngle = (i: number) => {
    if (i === 0) {
      return -90;
    }

    const copy = [...data];
    copy.splice(copy.length - i);

    const result = copy.reduce(
      (acc, cur) => acc + parseInt(cur[value] as string, 10),
      0
    );
    return -90 + 360 / (100 / result);
  };

  const renderCircle = (c: T, i: number) => {
    let lineAngle = calculateLineAngle(i);
    const percent = getPiePercent(c[value] as number);
    return (
      <G key={`circles-container-${i}`}>
        <Circle
          key={`c-${i}`}
          cx={centerX}
          cy={centerY}
          r={smallerRadius}
          fill={'transparent'}
          stroke={colors[i]}
          strokeWidth={radius}
          transform={`rotate(${lineAngle} ${centerX} ${centerY})`}
          strokeDasharray={`${percent} ${circumference}`}
        />
      </G>
    );
  };

  const render_circles = () => {
    return (
      <G>
        <Circle
          cx={centerX}
          cy={centerY}
          r={radius}
          strokeWidth={1}
          stroke={'#000'}
          fill={pieBackgroundColor}
        />
        {data.map((c, i) => renderCircle(c, i))}
      </G>
    );
  };

  const render_text = () => {
    return data.map((c, i) => {
      let lineAngle = calculateLineAngle(i);
      const percent = getPiePercent(c[value] as number);
      const fs = fontSize / 4;
      const textRotation = lineAngle + percent / 2 - fs;

      return (
        <G key={`text-${i}`}>
          <SvgText
            x={centerX + smallerRadius}
            y={centerY + fontSize / 2}
            fontSize={fontSize}
            fill={textColor}
            transform={`rotate(${textRotation} ${centerX} ${centerY})`}
          >
            {c[label] as React.ReactNode}
          </SvgText>
        </G>
      );
    });
  };

  const render_ticks = () => {
    return data.map((c, i) => {
      const lineAngle = calculateLineAngle(i);
      const percent = getPiePercent(c[value] as number);
      const lineRotation = lineAngle + percent / 2;
      const fs = fontSize / 8;
      const textRotation = lineAngle + percent / 2 - fs;
      return (
        <G key={`ticks-${i}`}>
          <Line
            key={`tick-${i}`}
            x1={centerX + radius}
            y1={centerY}
            x2={centerX + radius + 5}
            y2={centerY}
            stroke={tickColor}
            strokeWidth={1}
            transform={`rotate(${lineRotation} ${centerX} ${centerY})`}
          />

          <SvgText
            x={centerX + radius + 7}
            y={centerY + fontSize / 2}
            fill={tickTextColor}
            fontSize={fontSize}
            transform={`rotate(${textRotation} ${centerX} ${centerY})`}
          >
            {c[value] as React.ReactNode}%
          </SvgText>
        </G>
      );
    });
  };

  const render_donut = () => {
    return (
      <G key="donut">
        <Circle
          cx={centerX}
          cy={centerY}
          r={donutRadius}
          fill={donutColor}
          strokeWidth={1}
          stroke={'#000'}
        />
      </G>
    );
  };

  const chartContainer = {
    height: containerHeight,
    width: containerWidth,
    backgroundColor: chartBackgroundColor,
  };

  const svgContainer = {
    backgroundColor: 'transparent',
  };

  return (
    <View style={chartContainer}>
      <Svg height="100%" width="100%" style={svgContainer}>
        {render_circles()}
        {render_text()}
        {render_ticks()}
        {useDonut && render_donut()}
      </Svg>
    </View>
  );
};

export default PieChart;
