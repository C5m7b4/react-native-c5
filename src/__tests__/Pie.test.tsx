import React from 'react';
import renderer from 'react-test-renderer';
import PieChart from '../charts/PieChart';

const pieData = [
  { value: 50, label: 'Groc' },
  {
    value: 10,
    label: 'Meat',
  },
];

describe('PieChart', () => {
  test('should render correctly', () => {
    const tree = renderer
      .create(<PieChart data={pieData} value="value" label="label" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
