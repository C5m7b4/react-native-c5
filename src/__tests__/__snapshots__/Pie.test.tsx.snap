// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "transparent",
      "height": 300,
      "width": 700,
    }
  }
>
  <RNSVGSvgView
    bbHeight="100%"
    bbWidth="100%"
    focusable={false}
    height="100%"
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "backgroundColor": "transparent",
        },
        Object {
          "flex": 0,
          "height": "100%",
          "width": "100%",
        },
      ]
    }
    width="100%"
  >
    <RNSVGGroup
      fill={
        Object {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGGroup
        fill={
          Object {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGCircle
          cx="350"
          cy="150"
          fill={
            Object {
              "payload": 0,
              "type": 0,
            }
          }
          propList={
            Array [
              "fill",
              "stroke",
              "strokeWidth",
            ]
          }
          r="115"
          stroke={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
          strokeWidth="1"
        />
        <RNSVGGroup
          fill={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
        >
          <RNSVGCircle
            cx="350"
            cy="150"
            fill={
              Object {
                "payload": 0,
                "type": 0,
              }
            }
            matrix={
              Array [
                6.123233995736766e-17,
                -1,
                1,
                6.123233995736766e-17,
                199.99999999999997,
                500,
              ]
            }
            propList={
              Array [
                "fill",
                "stroke",
                "strokeWidth",
                "strokeDasharray",
              ]
            }
            r="57.5"
            stroke={
              Object {
                "payload": 4294752771,
                "type": 0,
              }
            }
            strokeDasharray={
              Array [
                "180.64157758141312",
                "361.28315516282623",
              ]
            }
            strokeWidth="115"
          />
        </RNSVGGroup>
        <RNSVGGroup
          fill={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
        >
          <RNSVGCircle
            cx="350"
            cy="150"
            fill={
              Object {
                "payload": 0,
                "type": 0,
              }
            }
            matrix={
              Array [
                6.123233995736766e-17,
                1,
                -1,
                6.123233995736766e-17,
                500,
                -200,
              ]
            }
            propList={
              Array [
                "fill",
                "stroke",
                "strokeWidth",
                "strokeDasharray",
              ]
            }
            r="57.5"
            stroke={
              Object {
                "payload": 4280787788,
                "type": 0,
              }
            }
            strokeDasharray={
              Array [
                "36.128315516282626",
                "361.28315516282623",
              ]
            }
            strokeWidth="115"
          />
        </RNSVGGroup>
      </RNSVGGroup>
      <RNSVGGroup
        fill={
          Object {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGText
          content={null}
          dx={Array []}
          dy={Array []}
          fill={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
          font={
            Object {
              "fontSize": "12",
            }
          }
          matrix={
            Array [
              0.9989069009783763,
              -0.046744017561356335,
              0.046744017561356335,
              0.9989069009783763,
              -6.629017976635112,
              16.524370999718258,
            ]
          }
          propList={
            Array [
              "fill",
            ]
          }
          rotate={Array []}
          x={
            Array [
              "407.5",
            ]
          }
          y={
            Array [
              "156",
            ]
          }
        >
          <RNSVGTSpan
            content="Groc"
            dx={Array []}
            dy={Array []}
            fill={
              Object {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
        </RNSVGText>
      </RNSVGGroup>
      <RNSVGGroup
        fill={
          Object {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGText
          content={null}
          dx={Array []}
          dy={Array []}
          fill={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
          font={
            Object {
              "fontSize": "12",
            }
          }
          matrix={
            Array [
              -0.2599004916966832,
              0.9656354044958285,
              -0.9656354044958285,
              -0.2599004916966832,
              585.8104827682134,
              -148.98731781903751,
            ]
          }
          propList={
            Array [
              "fill",
            ]
          }
          rotate={Array []}
          x={
            Array [
              "407.5",
            ]
          }
          y={
            Array [
              "156",
            ]
          }
        >
          <RNSVGTSpan
            content="Meat"
            dx={Array []}
            dy={Array []}
            fill={
              Object {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
        </RNSVGText>
      </RNSVGGroup>
      <RNSVGGroup
        fill={
          Object {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGLine
          fill={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
          matrix={
            Array [
              0.9999843266448797,
              0.005598791350522019,
              -0.005598791350522019,
              0.9999843266448797,
              0.8453043768704447,
              -1.957225969414651,
            ]
          }
          propList={
            Array [
              "stroke",
              "strokeWidth",
            ]
          }
          stroke={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
          strokeWidth="1"
          x1="465"
          x2="470"
          y1="150"
          y2="150"
        />
        <RNSVGText
          content={null}
          dx={Array []}
          dy={Array []}
          fill={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
          font={
            Object {
              "fontSize": "12",
            }
          }
          matrix={
            Array [
              0.9997882162630735,
              -0.020579665242707226,
              0.020579665242707226,
              0.9997882162630735,
              -3.0128254784817727,
              7.234650395486511,
            ]
          }
          propList={
            Array [
              "fill",
            ]
          }
          rotate={Array []}
          x={
            Array [
              "472",
            ]
          }
          y={
            Array [
              "156",
            ]
          }
        >
          <RNSVGTSpan
            content="50"
            dx={Array []}
            dy={Array []}
            fill={
              Object {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
          <RNSVGTSpan
            content="%"
            dx={Array []}
            dy={Array []}
            fill={
              Object {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
        </RNSVGText>
      </RNSVGGroup>
      <RNSVGGroup
        fill={
          Object {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGLine
          fill={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
          matrix={
            Array [
              -0.3100817593818749,
              0.9507098939732568,
              -0.9507098939732568,
              -0.3100817593818749,
              601.1350998796447,
              -136.23619898335863,
            ]
          }
          propList={
            Array [
              "stroke",
              "strokeWidth",
            ]
          }
          stroke={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
          strokeWidth="1"
          x1="465"
          x2="470"
          y1="150"
          y2="150"
        />
        <RNSVGText
          content={null}
          dx={Array []}
          dy={Array []}
          fill={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
          font={
            Object {
              "fontSize": "12",
            }
          }
          matrix={
            Array [
              -0.285088818357078,
              0.9585011036236551,
              -0.9585011036236551,
              -0.285088818357078,
              593.5562519685255,
              -142.7120635147176,
            ]
          }
          propList={
            Array [
              "fill",
            ]
          }
          rotate={Array []}
          x={
            Array [
              "472",
            ]
          }
          y={
            Array [
              "156",
            ]
          }
        >
          <RNSVGTSpan
            content="10"
            dx={Array []}
            dy={Array []}
            fill={
              Object {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
          <RNSVGTSpan
            content="%"
            dx={Array []}
            dy={Array []}
            fill={
              Object {
                "payload": 4278190080,
                "type": 0,
              }
            }
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
        </RNSVGText>
      </RNSVGGroup>
      <RNSVGGroup
        fill={
          Object {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGCircle
          cx="350"
          cy="150"
          fill={
            Object {
              "payload": 4294967295,
              "type": 0,
            }
          }
          propList={
            Array [
              "fill",
              "stroke",
              "strokeWidth",
            ]
          }
          r="50"
          stroke={
            Object {
              "payload": 4278190080,
              "type": 0,
            }
          }
          strokeWidth="1"
        />
      </RNSVGGroup>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
